src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs

```diff
@@ -48,85 +49,96 @@ public class PipelineArtifactServer
+                    {
+                        // each pipeline artifact will have its own subroot to avoid file collisions
+                        string pipelineArtifactRootPath = Path.Combine(downloadParameters.TargetDirectory, pipelineArtifact.Name);
+                        await DownloadPipelineArtifact(
+                            buildDropManager, 
+                            pipelineArtifact,
+                            pipelineArtifactRootPath, 
+                            downloadParameters.MinimatchFilters, 
+                            cancellationToken);
+                    }
+                }
             }
-            var manifestId = DedupIdentifier.Create(art.Resource.Data);
-
-            // 2) download to the target path
-            var buildDropManager = this.GetBDM(context, connection);
-            DownloadPipelineArtifactOptions options = DownloadPipelineArtifactOptions.CreateWithManifestId(manifestId, targetDir, proxyUri: null, minimatchPatterns: minimatchFilters);
-            await buildDropManager.DownloadAsync(options, cancellationToken);
-        }
-
-        // Download pipeline artifact with project name.
-        internal async Task DownloadAsyncWithProjectNameMiniMatch(
-            AgentTaskPluginExecutionContext context,
-            string project,
-            int buildId,
-            string artifactName,
-            string targetDir,
-            string[] minimatchFilters,
-            CancellationToken cancellationToken)
-        {
-            VssConnection connection = context.VssConnection;
-
-            // 1) get manifest id from artifact data
-            BuildServer buildHelper = new BuildServer(connection);
-            BuildArtifact art = await buildHelper.GetArtifactWithProjectAsync(project, buildId, artifactName, cancellationToken);
-            if (art.Resource.Type != "PipelineArtifact")
+            else
             {
-                throw new ArgumentException("The artifact is not of the type Pipeline Artifact\n");
-            }
-            var manifestId = DedupIdentifier.Create(art.Resource.Data);
+                // 1) get manifest id from artifact data
+                BuildArtifact buildArtifact;
+                if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectId)
+                {
+                    buildArtifact = await buildHelper.GetArtifact(downloadParameters.ProjectId, downloadParameters.BuildId, downloadParameters.ArtifactName, cancellationToken);
+                }
+                else if (downloadParameters.ProjectRetrievalOptions == BuildArtifactRetrievalOptions.RetrieveByProjectName)
+                {
+                    if (string.IsNullOrEmpty(downloadParameters.ProjectName))
+                    {
+                        throw new InvalidOperationException("Project name can't be empty when trying to fetch build artifacts!");
+                    }
+                    else
```

Comment:
Same here. The else statement is redundant.

Reply:
I agree explicit else better. Plus it forces the developer to think of all cases.

Reply:
The coding style is known as guard block and is common to be used to return/throw exception early if something went wrong (wrong parameter or invalid state). Adding a else statement will also work but it's redundant IMO.